// bisheng/lib/index.js
"use strict";

var _openBrowser = _interopRequireDefault(require("react-dev-utils/openBrowser"));

var _getWebpackCommonConfig = _interopRequireDefault(require("./config/getWebpackCommonConfig"));

var _updateWebpackConfig = _interopRequireDefault(require("./config/updateWebpackConfig"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var fs = require('fs');

var path = require('path');

var mkdirp = require('mkdirp');

var nunjucks = require('nunjucks');

var webpack = require('webpack');

var WebpackDevServer = require('webpack-dev-server');

var R = require('ramda');

var ghPages = require('gh-pages');

var _require = require('./utils/escape-win-path'),
    escapeWinPath = _require.escapeWinPath;

var getBishengConfig = require('./utils/get-bisheng-config');

var sourceData = require('./utils/source-data');

var generateFilesPath = require('./utils/generate-files-path');

var getThemeConfig = require('./utils/get-theme-config');

var context = require('./context');

var Module = require('module'); // We need to inject the require logic to support use origin node_modules
// if currently not provided.


var oriRequire = Module.prototype.require;

Module.prototype.require = function () {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  var moduleName = args[0];

  try {
    return oriRequire.apply(this, args);
  } catch (err) {
    var newArgs = args.concat();

    if (moduleName[0] !== '/') {
      newArgs[0] = path.join(process.cwd(), 'node_modules', moduleName);
    }

    return oriRequire.apply(this, newArgs);
  }
};

var tmpDirPath = path.join(__dirname, '..', 'tmp');
mkdirp.sync(tmpDirPath);

function getRoutesPath(themePath, configEntryName) {
  var routesTemplate = fs.readFileSync(path.join(__dirname, 'routes.nunjucks.js')).toString();
  var routesPath = path.join(tmpDirPath, "routes.".concat(configEntryName, ".js"));
  var bishengConfig = context.bishengConfig,
      themeConfig = context.themeConfig;
  fs.writeFileSync(routesPath, nunjucks.renderString(routesTemplate, {
    themePath: escapeWinPath(themePath),
    themeConfig: JSON.stringify(bishengConfig.themeConfig),
    themeRoutes: JSON.stringify(themeConfig.routes)
  }));
  return routesPath;
}

function generateEntryFile(configTheme, configEntryName, root) {
  var entryTemplate = fs.readFileSync(path.join(__dirname, 'entry.nunjucks.js')).toString();
  var entryPath = path.join(tmpDirPath, "entry.".concat(configEntryName, ".js"));
  var routesPath = getRoutesPath(path.dirname(configTheme), configEntryName);
  fs.writeFileSync(entryPath, nunjucks.renderString(entryTemplate, {
    routesPath: escapeWinPath(routesPath),
    root: escapeWinPath(root)
  }));
}

exports.start = function start(program) {
  var configFile = path.join(process.cwd(), program.config || 'bisheng.config.js');
  var bishengConfig = getBishengConfig(configFile);
  var themeConfig = getThemeConfig(bishengConfig.theme);
  context.initialize({
    bishengConfig: bishengConfig,
    themeConfig: themeConfig
  });
  mkdirp.sync(bishengConfig.output);
  var template = fs.readFileSync(bishengConfig.htmlTemplate).toString();
  var templateData = Object.assign({
    root: '/'
  }, bishengConfig.htmlTemplateExtraData || {});
  var templatePath = path.join(process.cwd(), bishengConfig.output, 'index.html');
  fs.writeFileSync(templatePath, nunjucks.renderString(template, templateData));
  generateEntryFile(bishengConfig.theme, bishengConfig.entryName, '/');
  var webpackConfig = (0, _updateWebpackConfig.default)((0, _getWebpackCommonConfig.default)(), 'start');
  webpackConfig.plugins.push(new webpack.HotModuleReplacementPlugin());

  var serverOptions = _objectSpread({
    quiet: true,
    hot: true
  }, bishengConfig.devServerConfig, {
    contentBase: path.join(process.cwd(), bishengConfig.output),
    historyApiFallback: true,
    host: 'localhost'
  });

  WebpackDevServer.addDevServerEntrypoints(webpackConfig, serverOptions);
  var compiler = webpack(webpackConfig); // Ref: https://github.com/pigcan/blog/issues/6
  // Webpack startup recompilation fix. Remove when @sokra fixes the bug.
  // https://github.com/webpack/webpack/issues/2983
  // https://github.com/webpack/watchpack/issues/25

  var timefix = 11000;
  compiler.plugin('watch-run', function (watching, callback) {
    watching.startTime += timefix;
    callback();
  });
  compiler.plugin('done', function (stats) {
    stats.startTime -= timefix;
  });
  var server = new WebpackDevServer(compiler, serverOptions);
  server.listen(bishengConfig.port, '0.0.0.0', function () {
    return (0, _openBrowser.default)("http://localhost:".concat(bishengConfig.port));
  });
};

var ssrTemplate = fs.readFileSync(path.join(__dirname, 'ssr.nunjucks.js')).toString();

function filenameToUrl(filename) {
  if (filename.endsWith('index.html')) {
    return filename.replace(/index\.html$/, '');
  }

  return filename.replace(/\.html$/, '');
}

exports.build = function build(program, callback) {
  var configFile = path.join(process.cwd(), program.config || 'bisheng.config.js');
  var bishengConfig = getBishengConfig(configFile);
  var themeConfig = getThemeConfig(bishengConfig.theme);
  context.initialize({
    bishengConfig: bishengConfig,
    themeConfig: themeConfig,
    isBuild: true
  });
  mkdirp.sync(bishengConfig.output);
  var entryName = bishengConfig.entryName;
  generateEntryFile(bishengConfig.theme, entryName, bishengConfig.root);
  var webpackConfig = (0, _updateWebpackConfig.default)((0, _getWebpackCommonConfig.default)(), 'build');
  webpackConfig.plugins.push(new webpack.LoaderOptionsPlugin({
    minimize: true
  }));
  webpackConfig.plugins.push(new webpack.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || 'production')
  }));
  var ssrWebpackConfig = Object.assign({}, webpackConfig);
  var ssrPath = path.join(tmpDirPath, "ssr.".concat(entryName, ".js"));
  var routesPath = getRoutesPath(path.dirname(bishengConfig.theme), entryName);
  fs.writeFileSync(ssrPath, nunjucks.renderString(ssrTemplate, {
    routesPath: escapeWinPath(routesPath)
  }));
  ssrWebpackConfig.entry = _defineProperty({}, "".concat(entryName, "-ssr"), ssrPath);
  ssrWebpackConfig.target = 'node';
  ssrWebpackConfig.output = Object.assign({}, ssrWebpackConfig.output, {
    path: tmpDirPath,
    library: 'ssr',
    libraryTarget: 'commonjs'
  });
  webpack(webpackConfig, function (err, stats) {
    if (err !== null) {
      return console.error(err);
    }

    if (stats.hasErrors()) {
      console.log(stats.toString('errors-only'));
      return;
    }

    var markdown = sourceData.generate(bishengConfig.source, bishengConfig.transformers);
    var filesNeedCreated = generateFilesPath(themeConfig.routes, markdown).map(bishengConfig.filePathMapper);
    filesNeedCreated = R.unnest(filesNeedCreated);
    var template = fs.readFileSync(bishengConfig.htmlTemplate).toString();

    if (!program.ssr) {
      require('./loaders/common/boss').jobDone();

      var templateData = Object.assign({
        root: bishengConfig.root
      }, bishengConfig.htmlTemplateExtraData || {});
      var fileContent = nunjucks.renderString(template, templateData);
      filesNeedCreated.forEach(function (file) {
        var output = path.join(bishengConfig.output, file);
        mkdirp.sync(path.dirname(output));
        fs.writeFileSync(output, fileContent);
        console.log('Created: ', output);
      });

      if (callback) {
        callback();
      }

      return;
    }

    context.turnOnSSRFlag(); // If we can build webpackConfig without errors, we can build ssrWebpackConfig without errors.
    // Because ssrWebpackConfig are just part of webpackConfig.

    webpack(ssrWebpackConfig, function (ssrBuildErr, ssrBuildStats) {
      if (ssrBuildErr) throw ssrBuildErr;
      if (ssrBuildStats.hasErrors()) throw ssrBuildStats.toString('errors-only');

      require('./loaders/common/boss').jobDone();

      var _require2 = require(path.join(tmpDirPath, "".concat(entryName, "-ssr"))),
          ssr = _require2.ssr;

      var fileCreatedPromises = filesNeedCreated.map(function (file) {
        var output = path.join(bishengConfig.output, file);
        mkdirp.sync(path.dirname(output));
        return new Promise(function (resolve) {
          ssr(filenameToUrl(file), function (error, content) {
            if (error) {
              console.error(error);
              process.exit(1);
            }

            var templateData = Object.assign({
              root: bishengConfig.root,
              content: content
            }, bishengConfig.htmlTemplateExtraData || {});
            var fileContent = nunjucks.renderString(template, templateData);
            fs.writeFileSync(output, fileContent);
            console.log('Created: ', output);
            resolve();
          });
        });
      });
      Promise.all(fileCreatedPromises).then(function () {
        if (callback) {
          callback();
        }
      });
    });
  });
};

function pushToGhPages(basePath, config) {
  var options = _objectSpread({}, config, {
    depth: 1,
    logger: function logger(message) {
      console.log(message);
    }
  });

  if (process.env.RUN_ENV_USER) {
    options.user = {
      name: process.env.RUN_ENV_USER,
      email: process.env.RUN_ENV_EMAIL
    };
  }

  ghPages.publish(basePath, options, function (err) {
    if (err) {
      throw err;
    }

    console.log('Site has been published!');
  });
}

exports.deploy = function deploy(program) {
  var config = {
    remote: program.remote,
    branch: program.branch
  };

  if (program.pushOnly) {
    var output = typeof program.pushOnly === 'string' ? program.pushOnly : './_site';
    var basePath = path.join(process.cwd(), output);
    pushToGhPages(basePath, config);
  } else {
    var configFile = path.join(process.cwd(), program.config || 'bisheng.config.js');
    var bishengConfig = getBishengConfig(configFile);

    var _basePath = path.join(process.cwd(), bishengConfig.output);

    exports.build(program, function () {
      return pushToGhPages(_basePath, config);
    });
  }
};
